chore(docs): Touch up profiler docs (#7524)
noir-lang/noir: docs/docs/tooling/profiler.md

---
title: Profiler
description: Learn about the Noir Profiler, how to generate execution flamegraphs, identify bottlenecks, and visualize optimizations.
keywords: [profiling, profiler, flamegraph]
sidebar_position: 3
---

The profiler is a sampling profiler designed to analyze and visualize Noir programs. It assists developers to identify bottlenecks by mapping execution data back to the original source code.

## Installation

The profiler is automatically installed with Nargo starting noirup v0.1.4.

Check if the profiler is already installed by running `noir-profiler --version`. If the profiler is not found, update noirup and install the profiler by [reinstalling both noirup and Nargo](../getting_started/quick_start.md#noir).

## Usage

### Profiling ACIR opcodes

The profiler provides the ability to flamegraph a Noir program's ACIR opcodes footprint. This is useful for _approximately_ identifying constrained execution and proving bottlenecks of Noir programs.

:::note

> "_Approximately_" as execution and proving speeds depend on the constrained execution trace and how the proving backend of choice interprets the ACIR opcodes.

Author: Do ACIR opcodes scale execution speeds "linearly" like proving backend gates scale proving speeds?

Maybe I was mistaken, but I thought ACIR opcodes could be heavier or lighter to execute depending on the exact opcode.

> "_Approximately_" as execution and proving speeds depend on the constrained execution trace and how the proving backend of choice interprets the ACIR opcodes.

Author: > isn't due to do "how the proving backend of choice interprets the ACIR opcodes"

Yup realized how confusing my wordings were when I'm re-reading it now.

Touched it up further at https://github.com/noir-lang/noir/pull/7524/commits/76c8dd9c3668c786907febabed1f944355cb800a, see if that's good?

> "_Approximately_" as execution and proving speeds depend on the constrained execution trace and how the proving backend of choice interprets the ACIR opcodes.

Author: > We could mention that certain ACIR opcodes can be heaveir but it felt like unnecessary extra info in this guide.

Staying on the more accurate side should be worthwhile, at least until we have more ready ACIR docs.

> "_Approximately_" as execution and proving speeds depend on the constrained execution trace and how the proving backend of choice interprets the ACIR opcodes.

Reviewer: Looks good!

    array
}
```

> Change directory into the project and compile the program using `nargo compile`. We are ready then to try the profiler out.

Reviewer: ```suggestion
Change into the project directory and compile the program using `nargo compile`. We are now ready to try out the profiler.
```
I think this reads better

> #### Flamegraphing

Reviewer: we should add this word to cspell

Let's take a granular look at our program's ACIR opcode footprint using the profiler, running:

```sh
noir-profiler opcodes --artifact-path ./target/program.json --output ./target/
```

> The command will generate a flamegraph in your _target_ folder that maps the number of ACIR opcodes and their corresponding locations in your program source code.

Reviewer: ```suggestion
The command generates a flamegraph in your _target_ folder that maps the number of ACIR opcodes to their corresponding locations in your program's source code.
```
For describing mappings, I find "to" clearer than "and"

Opening the flamegraph in a web browser will provide a more interactive experience, allowing you to click into different regions of the graph and examine them.

Flamegraph of the demonstrative project generated with Nargo v1.0.0-beta.2:

![ACIR Flamegraph Unoptimized](@site/static/img/tooling/profiler/acir-flamegraph-unoptimized.png)

> The demonstrative project consists of 387 ACIR opcodes in total, which from the flamegraph we can see that the majority of them comes from the write to `array[i]`.

Reviewer: ```suggestion
The demonstrative project consists of 387 ACIR opcodes in total. From the flamegraph, we can see that the majority come from the write to `array[i]`.
```
I see you want to add info about the total projects. I think this sentence becomes clearer when split.

> Knowing the insight on our program's bottleneck, let's optimize it.

Reviewer: ```suggestion
With insight into our program's bottleneck, let's optimize it.
```
"Knowing the insight" is a bit awkward

Flamegraph of the demonstrative project generated with bb v0.76.4:

![Gates Flamegraph Optimized](@site/static/img/tooling/profiler/gates-flamegraph-optimized.png)

The demonstrative project consists of 3,062 proving backend gates in total.

:::note

If you try searching for `i > ptr` in the source code, you will notice that this call stack is only contributing 3.8% of the total proving backend gates, versus the 43.3% ACIR opcodes it contributes.

This illustrates that number of ACIR opcodes are at best approximations of proving performances, where actual proving performances depend on how the proving backend interprets and translates ACIR opcodes into proving gates.

:::

#### Understanding bottlenecks

Profiling your program with different parameters is good way to understand your program's bottlenecks as it scales.

From the flamegraph above, you will notice that `blackbox::range` contributes the majority of the backend gates. This comes from how Barretenberg UltraHonk uses lookup tables for its range gates under the hood, which comes with a considerable but fixed setup cost in terms of proving gates.

> If our array is larger, range gates would become a much smaller percentage of our total circuit. See this flamegraph for the same program but with an array of size 2,048 (versus originally 32) in comparison:

Reviewer: let's note optimized

    array
}
```

> Instead of writing our array in a fully constrained manner, we first write our array inside an unconstrained function and then assert every value in the array returned from the unconstrained function in a constrained manner.

Reviewer: ```suggestion
Instead of writing our array in a fully constrained context, we first write our array inside an unconstrained function. Then, we assert every value in the array returned from the unconstrained function in a constrained context.
```

Reviewer: This sentence isn't a run-on, but I think it is a bit cleaner split up.

This brings the ACIR opcodes count of our program down to a total of 284 opcodes:

![ACIR Flamegraph Optimized](@site/static/img/tooling/profiler/acir-flamegraph-optimized.png)

#### Searching in flamegraphs

The `i > ptr` region in the above image is highlighted purple as we were searching for it.

Click "Search" in the top right corner of the flamegraph to start a search (e.g. i > ptr).

Check "Matched" in the bottom right corner to learn the percentage out of total opcodes associated with the search (e.g. 43.3%).

:::tip

If you try searching for `memory::op` before and after the optimization, you will find that the search will no longer have matches after the optimization.

> This comes from the optimization removing the use of a dynamic array (i.e. an array with a dynamic index, that is its values rely on witness inputs). After the optimized rewrite into reading two arrays from known constant indices, simple arithmetic operations replaces the original memory operations.

Reviewer: ```suggestion
This comes from the optimization removing the use of a dynamic array (i.e. an array with a dynamic index, that is its values rely on witness inputs). After the optimization, the program reads from two arrays with known constant indices, replacing the original memory operations with simple arithmetic operations.
```
Wdyt?

:::

> ### Profiling execution traces (unconstrained)

Reviewer: I do greatly prefer to keep the flow of ACIR opcodes -> backend gates -> Brillig flamegraphs.

I foresee most devs using the ACIR/backend gates profiling to reduce their circuit proving speeds over witness generation speeds. Backend gates also do not yet require introducing a `Prover.toml`, and `unconstrained`. Due to these reasons, I feel the backend gates section naturally follows the ACIR profiling section.

Author: I was debating if that's the right change as well.

I agree that devs' interests lie primarily with profiling backend gates; if anything, backend gates > execution > ACIR opcodes.

The question + my assumption that drove the change though was:

Do we intend to maintain backend gates profiling within noir-lang, and extend support for different backends?

I naively assumed no; but now that I think of it, maybe it's just a simple interface where backends will do the heavy lifting?

Author: The question + my assumption that drove the change though was:

Do we intend to maintain backend gates profiling within noir-lang, and extend support for different backends?

I naively assumed no; but now that I think of it, maybe it's just a simple interface where backends will do the heavy lifting?

Reviewer: > Do we intend to maintain backend gates profiling within noir-lang, and extend support for different backends?

Backends do the heavy lifting. The profiler simply accepts a path to a backend binary, the name of the their gates command, and allows to pass optional arguments (we do not document this here as it is not relevant for bb). Backends just need to follow our gates API if they want to integrate with the profiler.

Author: Lovely! I will update the ordering to Backend Gates > Brillig > ACIR in that case 👍 lmk if you have concerns.

Reviewer: I still prefer the ordering ACIR > Backend gates > Brillig as mentioned in the original comment. I think it flows better to keep the circuit flamegraphs together, as the backend flamegraph essentially builds upon the acir flamegraph (we still display acir opcodes in our backend flamegraph, just the sample count associated with those opcodes has now changed).

Author: Is profiling ACIR alone useful for devs?

If not, maybe we can also consider merging the ACIR and Backend Gates sections; wdyt

Reviewer: > Is profiling ACIR alone useful for devs?

Not sure what you mean by "merging" the sections. Do you mean removing the ACIR section? I do think they warrant separate sections as profiling ACIR is useful. Plus it is Noir's compilation target, so I think it should be documented how to profile it. 

It provides additional context as ACIR opcodes counts do not tie directly to backend gate counts. e.g. in this guide's example code backend gates from range opcodes take up most of the program, but it doesn't give us a full picture.
If we wanted to optimize witness generation we would need to look at the ACIR as well. Looking at the backend gates provides optimizations for proving speed. I realize this is why you wanted to go `Backend Gates > Brillig > ACIR`, but I think we should still go `ACIR > Backend gates > Brillig` for the reasons mentioned above.

```sh
noir-profiler execution-opcodes --artifact-path ./target/program.json --prover-toml-path Prover.toml --output ./target
```

This is similar to the `opcodes` command, except it additionally takes in the _Prover.toml_ file to profile execution with a specific set of inputs.

Flamegraph of the demonstrative project generated with Nargo v1.0.0-beta.2:
![Brillig Trace "Optimized"](@site/static/img/tooling/profiler/brillig-trace-opt-32.png)

Note that unconstrained Noir functions compile down to Brillig opcodes, which is what the counts in this flamegraph stand for, rather than constrained ACIR opcodes like in the previous section.

:::tip

Optimizing constrained operations through unconstrained rewrites like what we did in [the previous section](#optimizing-array-writes-with-reads) helps remove ACIR opcodes (hence shorter proving times), but would introduce more Brillig opcodes (hence longer execution times).

For example, we can find a 13.9% match `new_array` in the flamegraph above.

In contrast, if we profile the pre-optimization demonstrative project:
![Brillig Trace Initial Program](@site/static/img/tooling/profiler/brillig-trace-initial-32.png)

> You will notice that it does not consist any `new_array`, and executes a total of 1,582 Brillig opcodes (versus 2,125 Brillig opcodes post-optimization).

Reviewer: ```suggestion
You will notice that it does not contain `new_array` and executes a total of 1,582 Brillig opcodes (versus 2,125 Brillig opcodes post-optimization).
```

This feature depends on the proving backend you are using. We will use [Barretenberg](https://github.com/AztecProtocol/aztec-packages/tree/master/barretenberg) as an example here. Follow the [quick start guide](../getting_started/quick_start.md#proving-backend) to install it if you have not already.

#### Updating the demonstrative project

Let's first revert our demonstrative program into a constrained program by removing the `unconstrained` modifier from the main function:

```rust
fn main(...){...}
```

Then recompile the program running `nargo compile`.

#### Flamegraphing

Let's take a granular look at our program's proving backend gates footprint using the profiler, running:

```sh
noir-profiler gates --artifact-path ./target/program.json --backend-path bb --output ./target
```

The `--backend-path` flag takes in the path to your proving backend binary.

The above command assumes you have Barretenberg (bb) installed and that its path is saved in your PATH. If that is not the case, you can pass in the absolute path to your proving backend binary instead.

Flamegraph of the demonstrative project generated with bb v0.76.4:

> ![Gates Flamegraph Optimized](@site/static/img/tooling/profiler/gates-flamegraph-optimized.png)

Reviewer: Should we show the unoptimized one before this?

Author: Bumped the optimized flamegraph as this reads after the optimizing section.

Happy to reposition according to any reordering of sections from the parallel discussion above.
